### PR TITLE
Update GPUTreeshap

### PR DESCRIPTION
**Pull request by RAMitchell**
- Enable feature interactions

- Reduce python test time

- Add integration test with shap library

### CODE REVIEW

File: src/predictor/gpu_predictor.cu [hunk 1]
```diff
@@ -602,13 +602,74 @@ class GPUPredictor : public xgboost::Predictor {
...
-          phis.data().get() + batch.base_rowid * contributions_columns);
+          phis.data().get() + batch.base_rowid * contributions_columns, phis.size());
     }
     dh::safe_cuda(cudaMemcpy(contribs.data(), phis.data().get(),
                                   sizeof(float) * phis.size(),
                                   cudaMemcpyDefault));
   }
 
+  void PredictInteractionContributions(DMatrix* p_fmat,
+                                       std::vector<bst_float>* out_contribs,
```
Comment (trivialfis) [inline src/predictor/gpu_predictor.cu:613]: Should we change to host device vector instead?

File: src/predictor/gpu_predictor.cu [hunk 2]
```diff
@@ -602,13 +602,74 @@ class GPUPredictor : public xgboost::Predictor {
...
+                                       const gbm::GBTreeModel& model,
+                                       unsigned ntree_limit,
+                                       std::vector<bst_float>* tree_weights,
+                                       bool approximate) override {
+    if (approximate) {
```
Comment (trivialfis) [inline src/predictor/gpu_predictor.cu:618]: We need dispatching in gbm get predictor.

File: src/predictor/gpu_predictor.cu [hunk 3]
```diff
@@ -602,13 +602,74 @@ class GPUPredictor : public xgboost::Predictor {
...
+      batch.data.SetDevice(generic_param_->gpu_id);
+      batch.offset.SetDevice(generic_param_->gpu_id);
+      SparsePageView X(batch.data.DeviceSpan(), batch.offset.DeviceSpan(),
+                       model.learner_model_param->num_feature);
+      gpu_treeshap::GPUTreeShapInteractions(
+          X, device_paths.begin(), device_paths.end(), ngroup,
+          phis.data().get() + batch.base_rowid * contributions_columns, phis.size());
+    }
+    dh::safe_cuda(cudaMemcpy(out_contribs->data(), phis.data().get(),
+                                  sizeof(float) * phis.size(),
```
Comment (trivialfis) [inline src/predictor/gpu_predictor.cu:669]: Format.

File: tests/python/test_with_shap.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
+# -*- coding: utf-8 -*-
+import numpy as np
+import xgboost as xgb
+import testing as tm
+import unittest
+import pytest
+
+try:
```
Comment (trivialfis) [inline tests/python/test_with_shap.py:8]: Can we move this into testing.py?

File: tests/python/test_with_shap.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+# -*- coding: utf-8 -*-
```
Comment (trivialfis) [inline tests/python/test_with_shap.py:1]: Do we need this for Python 3?

File: tests/python/test_with_shap.py (hunk 4 above)
Comment (RAMitchell) [reply tests/python/test_with_shap.py:8]: I prefer it to stay here, I don't think it will be reused.

File: src/predictor/gpu_predictor.cu (hunk 2 above)
Comment (RAMitchell) [reply src/predictor/gpu_predictor.cu:618]: There might be of benefit to explicitly failing, maybe the user is expecting the algorithm to use GPU and it is silently switching to CPU.

File: src/predictor/gpu_predictor.cu (hunk 1 above)
Comment (RAMitchell) [reply src/predictor/gpu_predictor.cu]: Done

**Comment by trivialfis** [comment]
Can we close https://github.com/dmlc/xgboost/issues/5875 by this PR?

**Pull request merged**